File: checks/python/src/section508/checks/lists.py
from section508.dom import children_tags, closest, norm_text, select, tag_name


def check(root):
    if not root:
        return []
    if real_list(root):
        return []

    fakes = fake_list_items(root)
    if len(fakes) < 2:
        return []

    sample = norm_text(fakes[0])
    if sample:
        return [f"“{sample}” and the items next to it are not marked as a list."]
    return ["Related items are not marked as a list."]


def real_list(root):
    for lst in select(root, "ul, ol"):
        if any(tag_name(child) == "li" for child in children_tags(lst)):
            return True
    return any(select(lst, "dt, dd") for lst in select(root, "dl"))


def fake_list_items(root):
    items = []
    for element in select(root, "div, p"):
        if closest(element, "ul, ol, dl, li, dt, dd"):
            continue
        if select(element, "ul, ol, dl, p, div, table, h1, h2, h3, h4, h5, h6"):
            continue
        if not norm_text(element):
            continue
        items.append(element)

    by_parent = {}
    for element in items:
        parent = element.parent
        if parent is None:
            continue
        by_parent.setdefault(id(parent), []).append(element)

    for group in by_parent.values():
        if len(group) >= 2:
            return group
    return []

File: checks/ruby/lib/section508/checks/lists.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Related items are ul/ol/dl, not a stack of divs or paragraphs.
    class Lists
      def self.call(root)
        return [] unless root
        return [] if real_list?(root)

        fakes = fake_list_items(root)
        return [] if fakes.length < 2

        sample = Dom.norm(fakes.first)
        if sample.empty?
          ["Related items are not marked as a list."]
        else
          ["“#{sample}” and the items next to it are not marked as a list."]
        end
      end

      def self.real_list?(root)
        return false unless root.respond_to?(:css)

        lists = Dom.elements(root, "ul, ol")
        return true if lists.any? { |list| list.element_children.any? { |child| child.name.upcase == "LI" } }

        Dom.elements(root, "dl").any? { |list| list.at_css("dt, dd") }
      end

      def self.fake_list_items(root)
        return [] unless root.respond_to?(:css)

        items = Dom.elements(root, "div, p").select do |element|
          next false if Dom.closest(element, "ul, ol, dl, li, dt, dd")
          next false if element.at_css("ul, ol, dl, p, div, table, h1, h2, h3, h4, h5, h6")

          !Dom.text(element).strip.empty?
        end

        by_parent = items.group_by { |element| element.parent }
        by_parent.each_value do |group|
          return group if group.length >= 2
        end
        []
      end
    end
  end
end
